### ISSUE TITLE
Extraneous change events with Glimmer hash helper (performance regression)

### ISSUE THREAD

**Issue by kanongil** [emberjs/ember.js#14362]
When I observe a component property provided by a `hash` helper in Ember 2.9-beta.3 and canary, the observer fires any time the parent component is revalidated. It used to behave correctly in 2.8.

Failing pattern: ```{{my-component hash=(hash a=a)}}```

I have create a simple twiddle to demonstrate:

https://ember-twiddle.com/3d806dacb9ec04aa12afcdce0d6bc742?openFiles=components.my-component.js%2Ctemplates.components.my-component.hbs

In 2.8 the console output is:
```
```

In 2.9-beta.3 and canary it is:
```
aDidChange – "a"
```

**Comment by kanongil** [comment]
I updated the description to more accurately describe the issue.

Note that this issue causes my app to revalidate several hundred components that have not changed, effectively making it slower than 2.8.

**Comment by kanongil** [comment]
I tried to dig into this, and I think I discovered the root cause:

The hash helper is a plain `EvaluatedNamedArgs` instance. On invalidated `CurlyComponentManager.update()`, it retrieves all referenced properties and `setProperties()` on itself to update the component state. Since `EvaluatedNamedArgs#value()` is recomputed on each call, the `set()` call detects it as changed, and triggers observers on itself and child objects.

The fix should be relatively simple – either ensure that the object from `EvaluatedNamedArgs#value()` is the same on subsequent calls, or use an extended / wrapped version that does this from the hash helper.
